[pigeon] implemented primitive datatypes

gaaclarke opened the pull request:
fixes https://github.com/flutter/flutter/issues/66467

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

stuartmorgan commented on packages/pigeon/lib/java_generator.dart:
```diff
@@ -113,6 +113,19 @@ void _writeCodec(Indent indent, Api api) {
   });
 }
 
+String _boxedType(String type) {
```
Could you add a declaration comment here? It's non-obvious from the name that "boxing" a type that's not on the list is actually a no-op.

gaaclarke replied:
Added comments.

It was a mistake adding this method.  I started removing it in later PR's.  The way things are written there needs to to be a function like this that can return null and another that does a passthrough depending on the usage.  This should be joined with `_javaTypeForDartType` since it is duplication.

gaaclarke replied:
I removed this function completely and moved to _javaTypeForDartType in this PR: https://github.com/flutter/packages/pull/418

stuartmorgan commented on packages/pigeon/lib/java_generator.dart:
```diff
@@ -162,21 +176,25 @@ static MessageCodec<Object> getCodec() {
                   indent.writeln('@SuppressWarnings("ConstantConditions")');
-                  indent.writeln(
-                      '$argType input = $argType.fromMap((Map<String, Object>)message);');
+                  indent.writeln('$argType input = ($argType)message;');
```
I think I'm missing some high-level context here; why isn't the map lookup still required for other types?

gaaclarke replied:
Last week we moved to the codegen custom codec.  The codec will do the converting to/from a map for objects.  That opened the door to allow any argument that the codec supports to be sent.

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -92,8 +92,16 @@ const Map<String, String> _propertyTypeForDartTypeMap = <String, String>{
 
-String? _objcTypeForDartType(String type) {
-  return _objcTypeForDartTypeMap[type];
+String? _objcTypeForPrimitiveDartType(String type) {
```
It seems confusing that you are calling "Foo" the type in the map, but "Foo *" the type here, even though the usage is otherwise identical.

gaaclarke replied:
I agree.  I'm cleaning that a bit up in later PRs.  I renamed this `_objcTypePtrForPrimitiveDartType` to show we want a pointer type, not the type without the pointer.

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -92,8 +92,16 @@ const Map<String, String> _propertyTypeForDartTypeMap = <String, String>{
+      : null;
+}
+
+String _objcTypeForDartType(String? classPrefix, String type) {
```
Can you document this as well? I'm not sure what the class prefix is/is for.

gaaclarke replied:
Done.

stuartmorgan commented on packages/pigeon/lib/objc_generator.dart:
```diff
@@ -92,8 +92,16 @@ const Map<String, String> _propertyTypeForDartTypeMap = <String, String>{
+String _objcTypeForDartType(String? classPrefix, String type) {
+  return _objcTypeForDartTypeMap.containsKey(type)
+      ? _objcTypeForDartTypeMap[type]!
+      : _className(classPrefix, type);
```
You can write this without `!` (making it safer against future mistakes, such as if `type` were changed in one line but not the other) as:
```
final String? objcType = _objcTypeForDartTypeMap[type];
return objcType ?? _className(classPrefix, type);
```

gaaclarke replied:
I don't actually prefer this syntax since I think it hides the branch.  I didn't realize that until I already did it.  It's not a huge deal so I'll keep it changed like you mentioned.

stuartmorgan replied:
You could write it with an explicit branch if you prefer:
```
final String? objcType = _objcTypeForDartTypeMap[type];
if (objcType != null) {
   return objcType;
}
return _className(classPrefix, type);
```

But it is objectively better to write code that has enforced type safety, rather than code that bypasses type safety and relies on people to not make mistakes.

stuartmorgan commented on packages/pigeon/pigeons/primitive.dart:
```diff
@@ -0,0 +1,24 @@
+
+@FlutterApi()
+abstract class PrimitiveFlutterApi {
+  int inc(int value);
```
Why the strange abbreviation here?

gaaclarke replied:
TLDR Done

"inc" is a typical function found in other (old) languages that means "increment".  I was going to have this function do something but then decided that echoing is enough.

stuartmorgan commented on packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/all_datatypes.dart:
```diff
@@ -111,17 +111,16 @@ class HostEverything {
       );
     } else {
-      return Everything.decode(replyMap['result']!);
+      return (replyMap['result'] as Everything?)!;
```
As a follow-up: it would be helpful if these files were in a folder called "gen" or "generated", or had names including "generated". I didn't realize at first that I was reviewing a generated file because the comment at the top isn't in the review context.

stuartmorgan commented on packages/pigeon/platform_tests/ios_unit_tests/ios/RunnerTests/PrimitiveTest.m:
```diff
@@ -0,0 +1,29 @@
+      }];
+  [self waitForExpectations:@[ expectation ] timeout:1.0];
+}
+
```
Why aren't the other primitives tested, as they are on the Java side?

gaaclarke replied:
Test Driven Development, I just wrote the tests that would lead to the code changes I wanted.  I knew I didn't need them for objc because I was tapping into a function that was already written and tested.  On the Java side I was calling into a newly written `_boxedType` function which didn't have a full implementation (because it was a duplication of `_javaTypeForDartType` logic which was a mistake).

stuartmorgan replied:
> Test Driven Development, I just wrote the tests that would lead to the code changes I wanted.

This is the opposite of every description of TDD I've encountered. The standard TDD practice is that you write test cases for all of your requirements (which in this case is all of the primitive types working) and then write as little *implementation code* as possible to make those tests pass (which in this case happens to be trivial). Minimizing the amount of testing definitely isn't a TDD goal.

Starting with knowledge of the implementation internals to decide what tests you should write isn't actually test-driven, and the result is that your tests aren't actually asserting everything you want to be true. If someone later refactored in a way that it wasn't sharing that function any more, and broke all the other primitive types, no tests would fail, because you've never asserted that they work.

If Pigeon is supporting all the primitive types in all languages, then it should have tests covering all primitive types in all languages.

gaaclarke replied:
There is already another test for all the datatypes, it's called all_datatypes.  This was duplicating those tests and forging a duplicate path through the code with `_boxedType`, which is removed in a later PR.

The ideal for google's code is 100% test coverage.  By writing the test to assert the code before writing the code you are getting a higher test coverage.

stuartmorgan replied:
> The ideal for google's code is 100% test coverage.

If you mean LoC, that's a necessary (since any code that's not executed by tests isn't getting any coverage, obviously) but not sufficient condition for good testing.

This code has 100% test coverage by lines:

```
test_absolute_value() {
  assert(absolute_value(-1), 1);
}
```

```
int absolute_value(int value) {
  return value.abs();
}
```

I'm saying that if you start from the assumption that you are going to implement `absolute_value` by calling `abs`, and use that to justify only writing the one test above, that's not good test design, because:

```
int absolute_value(int value) {
  return value * -1;
}
```

still passes all tests (and even has 100% coverage!).

> By writing the test to assert the code before writing the code you are getting a higher test coverage.

I'm not sure what the relevance of this statement is to my comments. I did not suggest that one shouldn't write the tests before the code, I said one should write tests that assert all of the requirements, rather than simply writing the test to provide minimal coverage based on knowledge of the implementation.

> There is already another test for all the datatypes

A test for this specific code, or a test for code at another layer that this code is implemented in terms of? If the latter, see the example above; the existence of tests for `abs` doesn't guarantee correctness of `absolute_value`. If the former, then why does this new test exist at all?

gaaclarke replied:
Yea, I see what you are saying.  It's an interesting discussion and I see your point.  My goal is 100% test coverage, `value * -1` is fine if we only send in negative numbers.  The other side of the coin is building the tests like you are suggesting is making changes to the code more difficult, increasing engineering cost and maintenance cost.  Pigeon isn't in its lifecycle where that makes sense yet, to me.  We can chat about his on the phone later.

stuartmorgan replied:
> `value * -1` is fine if we only send in negative numbers.

The title of this PR is "implement primitive datatypes" though, not "implement support for 'int'".

gaaclarke replied:
Talked offline, it makes sense to add the tests like Android has.  FWIW they were getting tested that they compile, not tested that they execute.

gaaclarke replied:
Done, added objc tests.

gaaclarke commented:
@stuartmorgan I addressed the feedback.  I'm writing code so fast that some of this feedback would have been really helpful last week but now I started fixing it 3 commits later.  It makes things a bit harder to manage.

I wonder if it would be easier for you to just review the conglomeration of all the PRs?  I tried to divide the PRs by features, but since they build off of each other so much it might easier if you just see the end result.  Let me know what you think.

gaaclarke commented on packages/pigeon/platform_tests/flutter_null_safe_unit_tests/lib/null_safe_pigeon.dart:
```diff
@@ -202,7 +199,32 @@ class Api {
+    }
+  }
+
+  Future<int> inc(int arg) async {
```
It did actually get renamed https://github.com/flutter/packages/pull/414/files#diff-9e91b13fda610a88abcd78209d839aa424dfcda5975f24bb91449c60d8ce68a2R15

This is generated code I didn't check in (that we want to remove at some point but can't because of the analysis step).  I'll update the code.

gaaclarke replied:
Done, turns out there were actually 2 instances of 'inc'

stuartmorgan reviewed (approved):
LGTM, thanks for the updated tests!
